### PR TITLE
🚀feat: backend api for book reviews

### PR DESCRIPTION
**Pull request by kickthemoon0817**
## Summary

This PR implements the library review feature for the backend, enabling users to create, view, and like book reviews. It includes comprehensive API endpoints for managing user reviews, database models for storing review data, and complete test coverage for all functionality. The implementation follows RESTful principles and aligns with frontend expectations for seamless integration.

### Core Features
- **BookReview Model Updates**: Enhanced the `BookReview` model to support standalone reviews without requiring a linked `Book` object, allowing users to review any book by title and author
- **Review API Endpoints**: Implemented complete CRUD operations for book reviews
  - `GET /library/reviews/` - List all reviews by authenticated user
  - `POST /library/reviews/` - Create new book review
  - `POST /library/reviews/{id}/like/` - Toggle like/unlike on reviews
- **Serializers**: Created `ReviewSerializer` and `CreateReviewSerializer` with camelCase field naming to match frontend expectations
- **URL Routing**: Added URL patterns for library review endpoints in `backend/books/urls.py`

### Database Changes
- **Migration 0002**: Updated `BookReview` model schema to make `book` field optional and added `book_title` and `author_name` fields for standalone reviews
- **ReviewHelpfulVote Model**: Supports like/unlike functionality with proper user tracking

### Testing Infrastructure
- **Test Restructuring**: Reorganized books app tests into separate files for better maintainability
  - `test_views_reviews.py` - Tests for review CRUD operations (232 lines)
  - `test_views_likes.py` - Tests for like/unlike functionality (262 lines)
- **Coverage Configuration**: Updated pytest configuration for proper test coverage reporting
- **Test Runner Enhancements**: Added unit test and coverage options to `run_tests.py`

### Code Quality
- **Pre-commit Integration**: Merged `feat/precommit-format-lint` branch for consistent code formatting
- **EditorConfig & Ktlint**: Added configuration files for consistent coding styles across Python and Kotlin
- **Ruff Integration**: Configured Ruff for Python linting and formatting
- **Documentation**: Updated README with comprehensive development tools setup instructions

### Additional Improvements
- **macOS Compatibility**: Added `.DS_Store` to `.gitignore`
- **Development Tools**: Added formatter and linter scripts for both Python and Kotlin
- **Setup Scripts**: Created `setup_pre_commit.sh` for easy pre-commit hook installation

### Code Quality
- [x] Code follows project coding conventions (79 chars for Python)
- [x] Self-review completed
- [x] Code is well-commented (especially complex logic)
- [x] No new warnings or errors
- [x] Code formatted: `make format`
- [x] Linting passed: `make lint`

### Testing Requirements
- [x] Tests added/updated for new functionality
- [x] All tests pass locally: `make test`
- [ ] Test coverage ≥80% for new code
- [x] Integration tests added (if applicable)

### Documentation
- [x] Code documentation updated (docstrings, comments)
- [x] API documentation updated (if API changes)
- [x] README updated (if setup/usage changes)
- [ ] **Design Documentation updated on Wiki** (if architectural changes)
  - [ ] Revision history incremented
  - [ ] System architecture diagram updated (if applicable)
  - [ ] Class diagrams updated (frontend/backend, if applicable)
  - [ ] ER diagram updated (if database changes)
  - [ ] Testing plan updated (Iteration 3+)
  - Wiki: [Design Documentation](https://github.com/snuhcs-course/swpp-2025-project-team-10/wiki/Design-Documentation)

### Dependencies & Breaking Changes
- [x] No new dependencies (or justified below)
- [x] Dependencies pinned to specific versions
- [x] No breaking changes (or documented below)
- [x] Dependent changes merged and published

### SWPP Course Requirements
- [ ] Design Documentation updated for major changes
- [ ] Testing plan updated (Iteration 3+)
- [ ] User acceptance test stories unchanged (after Iteration 3)

## Related Issues

Closes #

## Screenshots / Demo

N/A - Backend API implementation

## Breaking Changes

None. This is a new feature addition with no impact on existing functionality.

### Implementation Highlights

1. **Flexible Review System**: The updated `BookReview` model allows users to review books without requiring them to exist in our database first. This provides better UX as users can review any book immediately.

2. **Frontend Alignment**: All serializers use camelCase field naming (e.g., `bookTitle`, `authorName`, `imageUrls`) to match frontend expectations, eliminating the need for field name transformation on the client side.

3. **Test Organization**: Tests are now organized by feature (reviews vs. likes) rather than in a single monolithic file, improving maintainability and making it easier to locate specific test cases.

4. **Development Workflow**: The integration of pre-commit hooks and formatters ensures consistent code quality across the team.

### Future Considerations

- Consider adding pagination for review lists when the number of reviews grows
- May want to add filtering/sorting options for reviews (by date, likes, etc.)
- Could implement review editing and deletion endpoints in future iterations

---

### Review Focus Areas

1. **Database Migration**: Please verify the migration `0002_alter_bookreview_unique_together_and_more.py` is safe and handles existing data correctly
2. **API Response Format**: Confirm the response format matches frontend expectations (especially camelCase field names)
3. **Test Coverage**: Review test cases for completeness, especially edge cases in like/unlike functionality
4. **Security**: Verify that users can only create reviews as themselves and cannot manipulate other users' reviews

### Review Checklist
- [ ] Code quality and conventions followed
- [ ] Tests are comprehensive and pass
- [ ] Documentation is clear and complete
- [ ] No security vulnerabilities
- [ ] Performance impact considered
- [ ] Design Documentation updated (if architectural changes)
- [ ] Changes align with project architecture
- [ ] Breaking changes properly documented

### CODE REVIEW

File: backend/books/urls.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+urlpatterns = [
+    # User's book reviews
+    path("reviews/", UserReviewListCreateView.as_view(), name="user-reviews"),
+    # Like/unlike a review
+    path(
+        "reviews/<int:pk>/like/",
+        ReviewLikeView.as_view(),
+        name="review-like",
+    ),
```
Comment (vnny27) [inline backend/books/urls.py:20]: 자기 리뷰에도 좋아요 기능이 있기는 했는데 우선은 피드에 좋아요 기능을 더하긴 했었어요. 이부분도 구현해야하는 부분이긴 하니까 다음 iteration에는 피드 좋아요부터 구현해야할 거 같아요

**Comment by vnny27** [comment]
프론트 api에 맞춰서 잘 구현해주신 거 같아 감사합니다. 바로 integration해봐도 괜찮아 보여요
